crate_universe/src/metadata.rs

```diff
@@ -80,6 +85,44 @@ impl MetadataGenerator for Generator {
     }
 }
 
+#[derive(Clone)]
+pub struct Cargo {
+    path: PathBuf,
+    full_version: Arc<Mutex<Option<String>>>,
```

Comment:
Why not just `Option<String>`?

Reply:
To do that, we'd need every method which takes a `Cargo` to take a mutable reference to it, which would require some futzing around. Instead, we can just pass around shared references, at the cost of acquiring a lock to check whether we've already looked up the full version.